File: examples/csharp-dotnet/subject_filtering/subscribe_subject_filter.cs
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */


/**
 * @file subscribe_subject_filter.cs
 *
 * This file contains a simple example outlining how to subscribe
 * to and retrieve messages from the middleware bus while using subject
 * exclusion filtering to reduce the scope of a wildcarded subject subscription.
 *
 * This example program is intended to be run before starting the
 * 'publish_subject_filter' example program.
 */
using GMSEC.API;
using GMSEC.API.UTIL;
using GMSEC.API.MIST;
using System;
using System.Threading;

class subscribe_subject_filter
{
	static private String EXAMPLE_SUBSCRIPTION_SUBJECT = "GMSEC.>";
	static private String FILTERED_MESSAGE_SUBJECT = "GMSEC.TEST.FILTERED";

	static int Main(string[] args)
	{
		if (args.Length < 1)
		{
			Console.WriteLine("usage subscribe_subject_filter.exe mw-id=<middleware ID>");
			return -1;
		}

		//o Load the command-line input into a GMSEC Config object
		// A Config object is basically a key-value pair map which is used to
		// pass configuration options into objects such as Connections,
		// ConnectionManagers, Subscribe and Publish function calls, Messages,
		// etc.
		Config config = new Config(args);

		// If it was not specified in the command-line arguments, set LOGLEVEL
		// to 'INFO' and LOGFILE to 'stdout' to allow the program report output
		// on the terminal/command line
		InitializeLogging(config);


		//o Print the GMSEC API version number using the GMSEC Logging
		// interface
		// This is useful for determining which version of the API is
		// configured within the environment
		Log.Info(ConnectionManager.GetAPIVersion());

		try
		{
			//o Create a ConnectionManager object
			// This is the linchpin for all communications between the
			// GMSEC API and the middleware server
			ConnectionManager connMgr = new ConnectionManager(config);

			//o Open the connection to the middleware
			Log.Info("Opening the connection to the middleware server");
			connMgr.Initialize();

			//o Output middleware client/wrapper version
			Log.Info(connMgr.GetLibraryVersion());

			//o Set up a subscription to listen for Messages with the topic
			// "GMSEC.TEST.PUBLISH" which are published to the middleware
			// bus
			// Subscription subject wildcard syntax:
			// * - Matches any one token separated by periods in a subject
			// > - Reading left to right, matches everything up to and ONE
			//     or more tokens in a subject
			// + - Reading left to right, matches everything up to and ZERO
			//     or more tokens in a subject
			// For more information on wildcards, please see the GMSEC API
			// User's Guide
			Log.Info("Subscribing to the topic: " + EXAMPLE_SUBSCRIPTION_SUBJECT);
			connMgr.Subscribe(EXAMPLE_SUBSCRIPTION_SUBJECT);

			//o Add a specific message subject to the Connection's exclusion
			// filter
			Log.Info("Adding a filter rule for the topic: " + FILTERED_MESSAGE_SUBJECT);
			connMgr.ExcludeSubject(FILTERED_MESSAGE_SUBJECT);

			//o Call receive() to synchronously retrieve a message that has
			// been received by the middleware client libraries
			// Timeout periods:
			// -1 - Wait forever
			//  0 - Return immediately
			// >0 - Time in milliseconds before timing out
			Log.Info("Waiting to receive a Message");
			Message message = connMgr.Receive(-1);

			// Example error handling for calling receive() with a timeout
			if (message != null)
			{
				Log.Info("Received message:\n" + message.ToXML());
			}

			Log.Info("Waiting 5 seconds to demonstrate that a second message will not be received");
			message = connMgr.Receive(5000);

			if (message != null)
			{
				Log.Error("Unexpectedly received a filtered message:\n" + message.ToXML());
			}

			//o Disconnect from the middleware and clean up the Connection
			connMgr.Cleanup();
		}
		catch (Exception e)
		{
			Log.Error(e.ToString());
			return -1;
		}

		return 0;
	}

	static void InitializeLogging(Config config)
	{
		String logLevel = config.GetValue("LOGLEVEL");
		String logFile = config.GetValue("LOGFILE");

		if (logLevel == null)
		{
			config.AddValue("LOGLEVEL", "INFO");
		}
		if (logFile == null)
		{
			config.AddValue("LOGFILE", "STDERR");
		}
	}
}

File: examples/nodejs/subscribe_subject_filter.js
#!/usr/bin/env node


/*
 * Copyright 2007-2024 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */


/**
 * subscribe_subject_filter.js
 *
 * This file contains a simple example outlining how to subscribe
 * to and retrieve messages from the middleware bus while using subject
 * exclusion filtering to reduce the scope of a wildcarded subject subscription.
 * 
 * This example program is intended to be run before starting the
 * 'publish_subject_filter' example program.
 */

const gmsec_api_home = process.env.GMSEC_API_HOME
const gmsec = require(gmsec_api_home + "/bin/gmsec_nodejs.node");

const utility = require("./utility.js");


class subscribe_subject_filter
{
	run()
	{
		try
		{
			// Set up connection configuration options provided by the user
			const config = utility.createConfig(process.argv);

			// Initialize log level for output
			utility.initializeLogLevel(config);

			// Validate messages before being sent to the bus
			config.addValue("gmsec-msg-content-validate-recv", "true");

			// Deduce subscription topic for the example program
			var all_messages_topic;
			var hb_messages_topic;
			const level = config.getIntegerValue("gmsec-schema-level", gmsec.Specification.SchemaLevel_LEVEL_0);

			switch (level)
			{
				case gmsec.Specification.SchemaLevel_LEVEL_0:
					all_messages_topic = "C2MS.>";
					hb_messages_topic  = "C2MS.*.*.*.*.*.MSG.HB.+";
					break;

				case gmsec.Specification.SchemaLevel_LEVEL_1:
					all_messages_topic = "C2MS-PIPE.>";
					hb_messages_topic  = "C2MS-PIPE.*.*.*.*.*.MSG.HB.+";
					break;

				case gmsec.Specification.SchemaLevel_LEVEL_2:
				default:
					all_messages_topic = "GMSEC.>";
					hb_messages_topic  = "GMSEC.*.*.MSG.HB.+";
					break;
				}

			// Create connection instance
			const conn = new gmsec.Connection(config);

			// Establish connection to the GMSEC Bus
			conn.connect();

			// Output information regarding the API version and connection
			gmsec.Log.info(gmsec.Connection.getAPIVersion());
			gmsec.Log.info("Middleware version = " + conn.getLibraryVersion());

			// Subscribe to receive all messages
			conn.setupSubscription(all_messages_topic);
			gmsec.Log.info("Subscribed to topic: " + all_messages_topic);

			// Add specific message topic to the Connection's exclusion filter
			conn.excludeSubject(hb_messages_topic);
			gmsec.Log.info("Filtering topic: " + hb_messages_topic);

			// Define timeout period, or use Gmsec.WAIT_FOREVER to block
			// indefinitely to receive a message
			let msg = conn.receive(gmsec.GMSEC_WAIT_FOREVER);

			gmsec.Log.info("Received LOG message:\n" + msg.toXML());

			// Acknowledge processing the message
			msg.acknowledge();

			// Although a second message (HB) is sent by the publisher, it will not
			// be delivered to this Connection object
			gmsec.Log.info("Waiting to receive other message(s)...");
			msg = conn.receive(5000);

			if (null == msg)
			{
				gmsec.Log.info("As expected, a timeout occurred (i.e. no HB message received)");
			}
			else
			{
				gmsec.Log.warning("Received unexpected message:\n" + msg.toXML());

				// Acknowledge processing the message
				msg.acknowledge();
			}

			// Disconnect from the GMSEC Bus
			conn.disconnect();
		}
		catch (e)
		{
			// If here, the most likely case is the middleware could not be
			// deduced, or a failure occurred when attempting to connect to
			// the GMSEC Bus
			gmsec.Log.error("Exception: " + e.message);
		}
	}
}


// Instantiate the subscribe_subject_filter example, and run it.
const example = new subscribe_subject_filter();
example.run();
